### DISCUSSION TITLE
Import Layout from DOCX file

### DISCUSSION THREAD

**Discussion by hvbargen** [eclipse-birt/birt#859] [Ideas]
Clients often want to "use MS Word for creating reports".
What most people do not understand is the complexity behind this.
Things start to get very complicated as soon as the report contains master-detail-relationships (in other words: loops), e.g. a an invoice and invoice items.

But nevertheless it could be very useful for rapid development if one could write kind of a template report (as a base for discussion and specification) and create an *.rptdesign file from it.

The intention is not to specify a full layout in MS word, only to speed up development. However, the experience of implementing this could possibly be used later for more advanced tools (e.g. a template engine).

What I think of is more straightforward.

Input: A Word document.
Output: An valid rptdesign file with a dummy datasource and dataset, the master page(s) and layout body is constructed from the word content, including limited support for formatting.

"Variables" in the Word document could be "mail merge fields" or maybe just words with a simple markup syntax (eg bash-like with ${name}.

The variables would be translated to BIRT data items or dynamic text items (configurable). The variable name would result in a column with the same name in the generated data set. The data set returns exactly one row.
The whole layout would be inside a list item bound to this data set.

To make this a usable report, it is (at least in principle) only necessary to replace the data source and data set with real ones and maybe add report parameters.
The tedious layout work is not strictly necessary, at least it would be reduced significantly depending on how good the layout import supports the formatting instructions.
The Word styles could be mapped to BIRT styles.

I wonder if anybody has ever tried such an approach.

A next step could be to make this even more useful by not using a dummy data source and data set, but to specify additional input, e.g. an existing *.rptdesign file containing application specific boilerplate code and layout and the name of an existing dataset in this report or in a *.rptlibrary.

**Comment by wimjongman** [comment]
This sounds like a customer that starts with BIRT, is overwhelmed with the designer, and then suggests this option.

**Comment by hvbtup** [comment]
Yes and No. 

Clients often wish to be able to customize their reports or to create their own reports.

Sometimes, in larger companies, they order BIRT courses. It is definitely possible to teach someone BIRT in ~3 days. But let's face it: You need developer blood in your veins to make full use of BIRT. And even then, it takes time - like every other piece of software, in particular development software. In most cases, the attendees just don't have the time to actually practice BIRT developing when they are back in their everyday business. And when you only work with BIRT say 2-3 hours per month, you won't make progress.
To really make full use of BIRT, you need the following skills: A very basic understanding of Javascript, a basic understanding of HTML, a good knowledge about SQL and your database system (assuming your reports use JDBC data sources), and last but not least a good knowledge of the application-specific data model. And you have to cope with the BIRT IDE of course.
In large teams, one might consider dividing into layout design specialits and data set design specialists, but our clients are not IT-centric companies, so those who are going to develop with BIRT are either "power-users" (which often lack basic knowledge) or IT people who have to handle a lot of very different topics and techniques. Neither is ideal.

But even if you have all the necessary skills and you know BIRT very good, it is still kind of tedious work to develop a layout with BIRT (the same holds true for other report development environments IMHO). And quite often, you cannot develop a layout directly in BIRT. Instead, someone else (e.g. the project lead) or the client him/herself wants to specify a layout. These people are used to MS office.

So you - as the developer - may receive a word document with the words "this is how the report should look like". Then you have to go and setup the master page, all the boilerplate text, grids and formatting. Then setup data source and data sets and link everything (often it saves some mouse clicks if you begin with the data sources and data sets and then drag and drop columns to the layout, of course).

It would save some boring work if at least the layout could be created more or less automatically from the BIRT document. 
For example, I know some developers who have difficulties with correct spelling or who use to work sloppy. For longer text passages, when they read and type text, this results in typos wihch can easily be go into production unseen :-). If the text would be automatically copied verbatim from the word document instead, there is one pitfall less (BTW: is that correct English?).

I estimate that such a tool could save 1/2 up to 3 hours development time per report - depending on the report, of course.

Another point is what I said in the beginning of the OP: Clients often demand on using MS office for reports, despite the fact that this is nonsense from a practical point of view, at least if you want a high degree of automation with fully auto-generated reports.

If - at least for simple reports without master-detail data models - they were able to use MS word to specify what field goes where etc, the workflow could be like this:

    Report X is to be generated for order no Y.
    There is a word document X.docx which serves as a template.
    If X.rptdesign does not exist or is older than X.docx:
         Create X.rptdesign from X.docx (this is the point of this discussion, assuming that a data source and data set is given)
    Create Y.pdf for order Y from X.rptdesign by calling BIRT.

If there are little changes in the layout (e.g. column widths, addditional boilerplate text and the like) this can be done by changing the word document only.
